src/UserCodeRunner.ts

```diff
@@ -1,482 +1,649 @@
-      identifier?: string,
-      cachedData?: Buffer | NodeJS.TypedArray | DataView,
-      context?: vm.Context,
-      lineOffset?: number,
-      columnOffset?: number,
-      initializeImportMeta?: {
-        meta?: any,
-        module?: vm.SourceTextModule,
-      },
-      importModuleDynamically?: (specifier: string, importMeta: any) => Promise<vm.Module>,
-    });
-    createCachedData(): Buffer;
-  }
+interface CacheItem {
+	jsFileMap: Map<string, ts.SourceFile>;
+	tsFileMap: Map<string, ts.SourceFile>;
+	sourceMap: BasicSourceMapConsumer | IndexedSourceMapConsumer;
+}
 
+export interface UserCodeRunnerOptions {
+	cache?: boolean;
+	cacheOptions?: LRUCache.Options<string, CacheItem>;
 }
 
-export async function executeUserCode<ArgsType extends any[], ReturnType = any>(
-  userCode: string,
-  args: ArgsType,
-  outputType: string = 'any',
-  argsTypes: string[] = ['any'],
-  timeout: number = 5000,
-  additionalSourceFiles: ts.SourceFile[] = [],
-  context: vm.Context = vm.createContext(),
-): Promise<Result<ReturnType, UserCodeError[]>> {
-
-  // Typecheck and transpile code
-  const userSourceFile = ts.createSourceFile(USER_FILE_ALIAS, userCode, ts.ScriptTarget.ESNext, undefined, ts.ScriptKind.TS);
-
-  const executionCode =  `
-    import defaultExport from '${USER_FILE_ALIAS}';
-    
-    declare global {
-      const args: [${argsTypes.join(', ')}];
-      let result: ${outputType};
-    }
-    result = defaultExport(...args);
-  `;
-  const executionSourceFile = ts.createSourceFile(EXECUTION_HARNESS_FILENAME, executionCode, ts.ScriptTarget.ESNext, undefined, ts.ScriptKind.TS);
-
-  const tsFileCache = new Map<string, ts.SourceFile>();
-
-  tsFileCache.set(`${USER_FILE_ALIAS}.ts`, userSourceFile);
-  tsFileCache.set(`${EXECUTION_HARNESS_FILENAME}.ts`, executionSourceFile);
-
-  for (const additionalSourceFile of additionalSourceFiles) {
-    tsFileCache.set(additionalSourceFile.fileName, additionalSourceFile);
-  }
-
-  const jsFileCache = new Map<string, ts.SourceFile>();
-
-  const defaultCompilerHost = ts.createCompilerHost({});
-  const customCompilerHost: ts.CompilerHost = {
-    ...defaultCompilerHost,
-    getCurrentDirectory(): string {
-      return '';
-    },
-    getSourceFile: (fileName, languageVersion) => {
-      if (tsFileCache.has(fileName)) {
-        return tsFileCache.get(fileName);
-      } else if (fileName.includes('typescript/lib')) {
-        return defaultCompilerHost.getSourceFile(fileName, languageVersion);
-      }
-      console.log(fileName);
-      return undefined;
-    },
-    writeFile: (filename, data) => {
-      jsFileCache.set(filename, ts.createSourceFile(filename, data, ts.ScriptTarget.ESNext, undefined, ts.ScriptKind.JS));
-    },
-    readFile(fileName: string): string | undefined {
-      if (tsFileCache.has(fileName)) {
-        return tsFileCache.get(fileName)!.text;
+export class UserCodeRunner {
+	private user_file_cache: LRUCache<string, CacheItem>;
```

Comment:
Looks like this file is using more spaces per tab than the other sources.

Reply:
☝️ this makes it quite difficult to read the diff in this file

Reply:
NVM, Github to the rescue:

<img width="302" alt="Screen Shot 2022-03-21 at 10 05 52 AM" src="https://user-images.githubusercontent.com/1189602/159326142-44c8afd1-c333-44c2-954d-ebeb84b6554f.png">